In [ ]:
import pandas as pd

# Load the dataset
df = pd.read_csv('WA_Fn-UseC_-Telco-Customer-Churn.csv')

# Basic shape
print("Rows, Columns:", df.shape)

# Column names and data types
print(df.info())

# First 5 rows
df.head()

Rows, Columns: (7043, 21)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,...,Yes,No,No,No,One year,No,Mailed check,56.95,1889.5,No
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,...,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes
3,7795-CFOCW,Male,0,No,No,45,No,No phone service,DSL,Yes,...,Yes,Yes,No,No,One year,No,Bank transfer (automatic),42.30,1840.75,No
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes


# Customer Churn Prediction — Week 1

## Problem Statement

Customer churn refers to when a customer stops using a company's product or
service — for example, cancelling a subscription or switching to a competitor.
Churn is one of the most important metrics for subscription-based and service
businesses (telecom, banking, insurance, e-commerce) because acquiring a new
customer is significantly more expensive than retaining an existing one. Even
a small increase in churn can quietly erode revenue over time.

The objective of this project is to build a Machine Learning classification
model that predicts whether a customer is likely to churn, based on their
demographic information, account details, subscription information, service
usage, contract type, tenure, and billing information.

By identifying customers who are at high risk of leaving, the company can
proactively intervene — through targeted offers, improved support, or
personalized retention strategies — before the customer actually cancels,
rather than reacting after the loss has already happened.

**Dataset:** Telco Customer Churn dataset (Kaggle)
**Target variable:** `Churn` (Yes / No)
**Problem type:** Binary Classification


## Data Dictionary

| Feature | Description | Type |
|---|---|---|
| customerID | Unique identifier for each customer | Categorical |
| gender | Customer's gender | Categorical |
| SeniorCitizen | Whether the customer is a senior citizen (1) or not (0) | Categorical |
| Partner | Whether the customer has a partner | Categorical |
| Dependents | Whether the customer has dependents | Categorical |
| tenure | Number of months the customer has stayed with the company | Numerical |
| PhoneService | Whether the customer has phone service | Categorical |
| MultipleLines | Whether the customer has multiple phone lines | Categorical |
| InternetService | Customer's internet service provider type (DSL/Fiber optic/No) | Categorical |
| OnlineSecurity | Whether the customer has online security add-on | Categorical |
| OnlineBackup | Whether the customer has online backup add-on | Categorical |
| DeviceProtection | Whether the customer has device protection add-on | Categorical |
| TechSupport | Whether the customer has tech support add-on | Categorical |
| StreamingTV | Whether the customer has streaming TV service | Categorical |
| StreamingMovies | Whether the customer has streaming movies service | Categorical |
| Contract | Type of contract (Month-to-month/One year/Two year) | Categorical |
| PaperlessBilling | Whether the customer uses paperless billing | Categorical |
| PaymentMethod | Customer's payment method | Categorical |
| MonthlyCharges | Amount charged to the customer monthly | Numerical |
| TotalCharges | Total amount charged to the customer | Numerical |
| Churn | Whether the customer left the company (target variable) | Target |

## Data Quality Report

**Dataset size:** 7,043 rows, 21 columns

**Missing Values:**
No missing values were detected by initial inspection (`df.info()` showed
7,043 non-null entries for every column). However, further investigation of
the `TotalCharges` column revealed 11 rows containing blank space characters
(`" "`) instead of true null values — these were not caught by standard
missing-value checks because Pandas read them as valid (but non-numeric) text.
All 11 affected rows had a `tenure` of 0, meaning these were new customers
with no billing history yet. These values were converted to numeric and
filled with 0.

**Duplicate Records:**
No duplicate rows and no duplicate `customerID` values were found.

**Incorrect Data Types:**
`TotalCharges` was originally stored as an `object` (text) column instead of
numeric, due to the hidden blank space characters described above. It was
converted to `float64` using `pd.to_numeric()`.

**Inconsistent Categorical Values:**
All categorical columns were checked for spelling errors, inconsistent
capitalization, and stray whitespace. No inconsistencies were found —
category values are clean and standardized across the dataset.

**Summary:**
The dataset was largely clean, with one notable issue in `TotalCharges` that
required correction. After cleaning, the dataset is ready for exploratory
data analysis and feature engineering in Week 2.

In [ ]:
# Try converting TotalCharges to numeric, and see what breaks
pd.to_numeric(df['TotalCharges'], errors='coerce').isna().sum()

np.int64(11)

In [ ]:
df[pd.to_numeric(df['TotalCharges'], errors='coerce').isna()][['tenure', 'MonthlyCharges', 'TotalCharges']]

,tenure,MonthlyCharges,TotalCharges
488,0,52.55,
753,0,20.25,
936,0,80.85,
1082,0,25.75,
1340,0,56.05,
3331,0,19.85,
3826,0,25.35,
4380,0,20.00,
5218,0,19.70,
6670,0,73.35,


In [ ]:
# Convert to numeric for real, turning the blanks into NaN
df['TotalCharges'] = pd.to_numeric(df['TotalCharges'], errors='coerce')

# Since these are new customers with no billing history yet, 0 makes sense
df['TotalCharges'] = df['TotalCharges'].fillna(0)

# Confirm it's now a proper number type
df['TotalCharges'].dtype

dtype('float64')

In [ ]:
print("Duplicate rows:", df.duplicated().sum())
print("Duplicate customer IDs:", df['customerID'].duplicated().sum())

Duplicate rows: 0
Duplicate customer IDs: 0


In [ ]:
# Check unique values in each categorical (object) column
for col in df.select_dtypes(include='object').columns:
    print(col, ":", df[col].unique())

customerID : ['7590-VHVEG' '5575-GNVDE' '3668-QPYBK' ... '4801-JZAZL' '8361-LTMKD'
 '3186-AJIEK']
gender : ['Female' 'Male']
Partner : ['Yes' 'No']
Dependents : ['No' 'Yes']
PhoneService : ['No' 'Yes']
MultipleLines : ['No phone service' 'No' 'Yes']
InternetService : ['DSL' 'Fiber optic' 'No']
OnlineSecurity : ['No' 'Yes' 'No internet service']
OnlineBackup : ['Yes' 'No' 'No internet service']
DeviceProtection : ['No' 'Yes' 'No internet service']
TechSupport : ['No' 'Yes' 'No internet service']
StreamingTV : ['No' 'Yes' 'No internet service']
StreamingMovies : ['No' 'Yes' 'No internet service']
Contract : ['Month-to-month' 'One year' 'Two year']
PaperlessBilling : ['Yes' 'No']
PaymentMethod : ['Electronic check' 'Mailed check' 'Bank transfer (automatic)'
 'Credit card (automatic)']
Churn : ['No' 'Yes']


In [ ]:
# Count and percentage of churned vs retained customers
print(df['Churn'].value_counts())
print(df['Churn'].value_counts(normalize=True) * 100)

Churn
No     5174
Yes    1869
Name: count, dtype: int64
Churn
No     73.463013
Yes    26.536987
Name: proportion, dtype: float64


In [ ]:
df.to_csv('cleaned_telco_churn.csv', index=False)